// ngx-datatable-bootstrap4/esm5/components/types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function DataTableTranslations() { }
function DataTableTranslations_tsickle_Closure_declarations() {
    /** @type {?} */
    DataTableTranslations.prototype.indexColumn;
    /** @type {?} */
    DataTableTranslations.prototype.selectColumn;
    /** @type {?} */
    DataTableTranslations.prototype.expandColumn;
    /** @type {?} */
    DataTableTranslations.prototype.paginationLimit;
    /** @type {?} */
    DataTableTranslations.prototype.paginationRange;
}
export var /** @type {?} */ defaultTranslations = /** @type {?} */ ({
    indexColumn: 'index',
    selectColumn: 'select',
    expandColumn: 'expand',
    paginationLimit: 'Limit',
    paginationRange: 'Results'
});
/**
 * @record
 */
export function DataTableParams() { }
function DataTableParams_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    DataTableParams.prototype.offset;
    /** @type {?|undefined} */
    DataTableParams.prototype.limit;
    /** @type {?|undefined} */
    DataTableParams.prototype.sortBy;
    /** @type {?|undefined} */
    DataTableParams.prototype.sortAsc;
}
/**
 * @record
 * @template T
 */
export function DataTableRowEvent() { }
function DataTableRowEvent_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    DataTableRowEvent.prototype.row;
    /** @type {?|undefined} */
    DataTableRowEvent.prototype.event;
}
/**
 * @record
 * @template T
 */
export function DataTableCellEvent() { }
function DataTableCellEvent_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    DataTableCellEvent.prototype.row;
    /** @type {?|undefined} */
    DataTableCellEvent.prototype.column;
    /** @type {?|undefined} */
    DataTableCellEvent.prototype.event;
}
/**
 * @record
 * @template T
 */
export function DataTableHeaderEvent() { }
function DataTableHeaderEvent_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    DataTableHeaderEvent.prototype.column;
    /** @type {?|undefined} */
    DataTableHeaderEvent.prototype.event;
}
