// ngx-datatable-bootstrap4/esm2015/components/title/title.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, forwardRef, HostListener, Inject } from '@angular/core';
import { DataTable } from '../../';
/**
 * @template T
 */
export class DataTableTitleComponent {
    /**
     * @param {?} dataTable
     */
    constructor(dataTable) {
        this.dataTable = dataTable;
        this.columnSelectorOpen = false;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    _closeSelector($event) {
        this.columnSelectorOpen = false;
    }
}
DataTableTitleComponent.decorators = [
    { type: Component, args: [{
                selector: 'data-table-title',
                template: `<div class="navbar">
  <h4 class="title" [textContent]="dataTable.title.title"></h4>
  <ng-container *ngIf="dataTable.title.titleTemplate" [ngTemplateOutlet]="dataTable.title.titleTemplate"
                [ngTemplateOutletContext]="{items: dataTable.items}">
  </ng-container>
  <div *ngIf="dataTable.title.controls">
    <button type="button" class="btn btn-default btn-sm refresh-button" (click)="dataTable.reloadItems()">
      <i class="fa fa-refresh"></i>
    </button>
    <button type="button" class="btn btn-default btn-sm column-selector-button" [class.active]="columnSelectorOpen"
            (click)="columnSelectorOpen = !columnSelectorOpen; $event.stopPropagation()">
      <i class="fa fa-list"></i>
    </button>
    <div class="column-selector-wrapper" (click)="$event.stopPropagation()">
      <div *ngIf="columnSelectorOpen" class="column-selector-box card">
        <div *ngIf="dataTable.expandableRows" class="column-selector-fixed-column checkbox">
          <label>
            <input type="checkbox" [(ngModel)]="dataTable.expandColumnVisible"/>
            <span>{{dataTable.translations.expandColumn}}</span>
          </label>
        </div>
        <div *ngIf="dataTable.indexColumn" class="column-selector-fixed-column checkbox">
          <label>
            <input type="checkbox" [(ngModel)]="dataTable.indexColumnVisible"/>
            <span>{{dataTable.translations.indexColumn}}</span>
          </label>
        </div>
        <div *ngIf="dataTable.selectColumn" class="column-selector-fixed-column checkbox">
          <label>
            <input type="checkbox" [(ngModel)]="dataTable.selectColumnVisible"/>
            <span>{{dataTable.translations.selectColumn}}</span>
          </label>
        </div>
        <div *ngFor="let column of dataTable.columns" class="column-selector-column checkbox">
          <label>
            <input type="checkbox" [(ngModel)]="column.visible"/>
            <span [textContent]="column.header"></span>
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`,
                styles: [`.navbar{padding:10px 0}.navbar .title{margin:5px 0 0 5px}.column-selector-wrapper{position:relative}.column-selector-wrapper .column-selector-box{box-shadow:0 0 10px #d3d3d3;width:150px;padding:10px;position:absolute;right:0;top:1px;z-index:1060}.column-selector-wrapper .column-selector-box .checkbox{margin-bottom:4px}.column-selector-wrapper .column-selector-box .column-selector-fixed-column{font-style:italic}`]
            },] },
];
/** @nocollapse */
DataTableTitleComponent.ctorParameters = () => [
    { type: DataTable, decorators: [{ type: Inject, args: [forwardRef(() => DataTable),] }] }
];
DataTableTitleComponent.propDecorators = {
    _closeSelector: [{ type: HostListener, args: ['document:click', ['$event'],] }]
};
function DataTableTitleComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DataTableTitleComponent.prototype.columnSelectorOpen;
    /** @type {?} */
    DataTableTitleComponent.prototype.dataTable;
}
