// ngx-datatable-bootstrap4/esm2015/components/pagination/pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, forwardRef, Inject } from '@angular/core';
import { DataTable } from '../../';
/**
 * @template T
 */
export class DataTablePagination {
    /**
     * @param {?} dataTable
     */
    constructor(dataTable) {
        this.dataTable = dataTable;
    }
    /**
     * @return {?}
     */
    get maxPage() {
        return Math.ceil(this.dataTable.itemCount / this.dataTable.limit);
    }
    /**
     * @return {?}
     */
    get limit() {
        return this.dataTable.limit;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set limit(value) {
        if (Number(value) > 0) {
            this.dataTable.limit = Math.floor(+value);
        }
    }
    /**
     * @return {?}
     */
    get page() {
        return this.dataTable.page;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set page(value) {
        if (Number(value) > 0) {
            this.dataTable.page = Math.floor(+value);
        }
    }
    /**
     * @return {?}
     */
    pageBack() {
        this.dataTable.offset -= Math.min(this.dataTable.limit, this.dataTable.offset);
    }
    /**
     * @return {?}
     */
    pageForward() {
        this.dataTable.offset += this.dataTable.limit;
    }
    /**
     * @return {?}
     */
    pageFirst() {
        this.dataTable.offset = 0;
    }
    /**
     * @return {?}
     */
    pageLast() {
        this.dataTable.offset = (this.maxPage - 1) * this.dataTable.limit;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    keyPress(event) {
        let /** @type {?} */ inputChar = String.fromCharCode(event.charCode);
        if (!(Number(inputChar) >= 0)) {
            event.preventDefault();
        }
    }
}
DataTablePagination.decorators = [
    { type: Component, args: [{
                selector: 'data-table-pagination',
                template: `<div class="navbar p-0 pagination-component">
  <div>
    {{dataTable.translations.paginationRange}}:&nbsp;{{dataTable.offset < 0 ? 0 : dataTable.offset + 1}}&nbsp;-
    {{[dataTable.offset + dataTable.limit , dataTable.itemCount] | min}}&nbsp;/&nbsp;{{dataTable.itemCount}}
  </div>
  <ul class="nav justify-content-end">
    <li class="nav-item pagination-limit">
      <div class="input-group">
        <div class="input-group-prepend">
          <span class="input-group-text">{{dataTable.translations.paginationLimit}}:</span>
        </div>
        <input #limitInput type="number" class="form-control" min="1" step="1"
               [ngModel]="limit" (blur)="limit = limitInput.value"
               (keyup.enter)="limit = limitInput.value"
               (keyup.esc)="limitInput.value = limit"
               (keypress)="keyPress($event)"/>
      </div>
    </li>
    <li class="nav-item">
      <button [disabled]="dataTable.offset <= 0" (click)="pageFirst()" class="btn btn-default">&laquo;</button>
      <button [disabled]="dataTable.offset <= 0" (click)="pageBack()" class="btn btn-default">&lsaquo;</button>
      <div class="d-inline-block">
        <div class="input-group">
          <input #pageInput type="number" class="form-control" min="1" step="1" max="{{maxPage}}"
                 [ngModel]="page" (blur)="page = pageInput.value"
                 (keyup.enter)="page = pageInput.value" (keyup.esc)="pageInput.value = page"
                 (keypress)="keyPress($event)"/>
          <div class="input-group-append">
            <span class="input-group-text">/&nbsp;{{dataTable.lastPage}}</span>
          </div>
        </div>
      </div>
      <button [disabled]="(dataTable.offset + dataTable.limit) >= dataTable.itemCount" (click)="pageForward()"
              class="btn btn-default">&rsaquo;
      </button>
      <button [disabled]="(dataTable.offset + dataTable.limit) >= dataTable.itemCount" (click)="pageLast()"
              class="btn btn-default">&raquo;
      </button>
    </li>
  </ul>
</div>
`,
                styles: [`.pagination-component .pagination-limit{margin-right:1rem}.pagination-component .pagination-limit input{width:75px}.pagination-component .btn{margin-top:-3px}`]
            },] },
];
/** @nocollapse */
DataTablePagination.ctorParameters = () => [
    { type: DataTable, decorators: [{ type: Inject, args: [forwardRef(() => DataTable),] }] }
];
function DataTablePagination_tsickle_Closure_declarations() {
    /** @type {?} */
    DataTablePagination.prototype.dataTable;
}
